fvm/env.go

```diff
@@ -909,3 +961,228 @@ func (e *transactionEnv) UpdateAccountContractCode(address runtime.Address, name
 func (e *transactionEnv) RemoveAccountContractCode(address runtime.Address, name string) (err error) {
 	return e.contracts.RemoveContract(address, name, e.GetSigningAccounts())
 }
+
+// AddAccountKey adds a public key to an existing account.
+//
+// This function returns an error if the specified account does not exist or
+// if the key insertion fails.
+func (e *transactionEnv) AddAccountKey(address runtime.Address,
+	publicKey *runtime.PublicKey,
+	hashAlgo runtime.HashAlgorithm,
+	weight int,
+) (
+	*runtime.AccountKey,
+	error,
+) {
+	accountAddress := flow.Address(address)
+
+	ok, err := e.accounts.Exists(accountAddress)
+	if err != nil {
+		// TODO: improve error passing https://github.com/onflow/cadence/issues/202
+		return nil, err
+	}
+
+	if !ok {
+		// TODO: improve error passing https://github.com/onflow/cadence/issues/202
+		return nil, fmt.Errorf("account with address %s does not exist", address)
+	}
+
+	signAlgorithm := RuntimeToCryptoSigningAlgorithm(publicKey.SignAlgo)
+	if signAlgorithm == crypto.UnknownSignatureAlgorithm {
+		// TODO: improve error passing https://github.com/onflow/cadence/issues/202
+		return nil, fmt.Errorf("cannot decode public key: invalid signature algorithm: %s", publicKey.SignAlgo)
+	}
+
+	hashAlgorithm := RuntimeToCryptoHashingAlgorithm(hashAlgo)
+	if hashAlgorithm == crypto.UnknownHashAlgorithm {
+		// TODO: improve error passing https://github.com/onflow/cadence/issues/202
+		return nil, fmt.Errorf("cannot decode public key: invalid hash algorithm: %s", hashAlgo)
+	}
+
+	decodedPublicKey, err := crypto.DecodePublicKey(signAlgorithm, publicKey.PublicKey)
+	if err != nil {
+		// TODO: improve error passing https://github.com/onflow/cadence/issues/202
+		return nil, fmt.Errorf("cannot decode public key: %w", err)
+	}
+
+	keyIndex, err := e.accounts.GetPublicKeyCount(accountAddress)
+	if err != nil {
+		// TODO: improve error passing https://github.com/onflow/cadence/issues/202
+		return nil, err
+	}
+
+	accountPublicKey := flow.AccountPublicKey{
+		Index:     int(keyIndex),
+		PublicKey: decodedPublicKey,
+		SignAlgo:  signAlgorithm,
+		HashAlgo:  hashAlgorithm,
+		SeqNumber: 0,
+		Weight:    weight,
+		Revoked:   false,
+	}
+
+	err = e.accounts.AppendPublicKey(accountAddress, accountPublicKey)
```

Comment:
I think we don't check that the hashing algorithm and the signature algorithm are supported for the Flow account model.
This can be solved by only supporting `SHA2-256`, `SHA3-256`, `ECDSA_P256` and `ECDSA_secp256k1` in Cadence for now. However we will hit the same issue when Cadence supports more algorithms than the Flow key model, so I think it's better to add a specific check in the this function before appending new keys.

Reply:
I'm not sure if the same function is called for account creation. If not the same algorithms check should be called during the account creation.

Reply:
Sounds good, I'll explicitly check that these hashing/signing algorithms.

There's no key addition during account creation.

Reply:
Added in a4b13ca60fac79ebd21ed4515ba4f8d9459e533f. Could you please have a look @tarakby?

Reply:
Thank you! that looks great 👌